Sources/Basic/JSON.swift

```diff
@@ -39,10 +39,17 @@ public enum JSON {
     case string(String)
 
     /// An array.
-    case array([JSON])
+    case array([JSONConvertible])
 
     /// A dictionary.
-    case dictionary([String: JSON])
+    case dictionary([String: JSONConvertible])
+}
+
+/// A JSON representation of an element.
+public protocol JSONConvertible {
+    
+    /// Return a JSON representation.
+    var json: JSON { get }
```

Comment:
I actually preferred this as a function -- I don't think properties should be used when they aren't close to O(1), and I would expect most implementations to just perform the transformation.